### PR TITLE
Feature/apm 303449 end to end automated cloud test for azure logs forwarder forwarder deployment

### PR DESCRIPTION
**Pull request by Dawid-Kaszubski**
E2E Test Flow:
_Before_
1. Build zip
2. Login to Azure with service principal credentials
3. Upload arm and zipped files to StorageAccount
_Main_
4. Run deployment script with StorageAccount URL repository
_After_
5. Remove files form StorageAccount
6. Run empty ARM deployment to remove all resources from ResourceGroup
7. Remove deployments

### CODE REVIEW

**Review by mswiatkowska — DISMISSED** [review]

**Review by pgutkowski — APPROVED** [review]

File: .travis.yml [hunk 1]
```diff
@@ -29,6 +30,42 @@ jobs:
...
+        - secure: HGKRJnJ2Y/JURFKWKWNtFpOMk1u99WIest+2SbuDBlDSVEtVWmUjUpAjz+TdaFt1XIaTANiQrHJdmsc4qgQW4QOhqjpOa4pm3CY74OfG7BYsBsaE5Ja8AiCYoiJ57VFejNQQdUMAzV8GWBjpX0CHbSwsG38ObyU6msql+DhLQXtyZZgbVXHq1DwrCAFB3xDazzePzcRuVikIZ70RoM5qa7VLxfWhI1oqHp31il4x3lr+vPYjWDqNulzXXv+8mD9A/GpEtAlDRJ2u2kH/3lSVNkIncUGKbb4Z2OiNUA97V7COLa31lhuDGSTAxVbf9xM9nnEHz/np3z9IHBZMvlCOCIBSPwgClCkUqEp0efKt6OufuAV/TkyFi0Llp4jeaTGeRRfPYF6hAZBbAR+2mOlyom6L6MBw0jrdnJq7euqdS/k/CdxkyMaXGGjqridGRrsRwZ//2ULcXbMgmKehSaqez/1ss4UtnV6LQqqbQQbZQppywz/j9bUoS8XGxaOAfhATRxOLbqQgKYmRD5YnJB+leVwFg3/YpTAiReEMrSwbyEN8OcAenHZtdCg+JwV/ZgobcerBTK0vXcWWlQgttqQEyT41qRU5kvkDfCc1UFDYWkVfaaRs0t+l4W+2YmtzIpDQk081JqFZHz6fXbadAIQu+J8Igap6bp9Y86eMUlLIwc4=
+      install:
+        - curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash
+      before_script:
+        - az login --service-principal --username "$USER_NAME"  --password "$PASSWORD" --tenant "$TENANT" > /dev/null
+        - chmod +x buildZip.sh
```
Comment (karolkieglerski) [inline .travis.yml:55]: could you add that `X` (execute flag) to this files before commit? Then we will not need that lines 55, 56 and 63

File: .travis.yml [hunk 2]
```diff
@@ -29,6 +30,42 @@ jobs:
...
+        - chmod +x version.sh
+        - ./buildZip.sh
+        - mv publish.zip dynatrace-azure-log-forwarder.zip
```
Comment (karolkieglerski) [inline .travis.yml:58]: it is really needed? if you adding `--path "dynatrace-azure-log-forwarder.zip"` it should also change a name of file

File: .travis.yml (hunk 1 above)
Comment (Dawid-Kaszubski) [reply .travis.yml]: WOW, I wasn't aware it's possible. Thanks!

**Review by mswiatkowska — APPROVED** [review]

**Review by karolkieglerski — APPROVED** [review]

**Review by pgutkowski — DISMISSED** [review]

**Review by sq2gxo — APPROVED** [review]

File: .travis.yml [hunk 3]
```diff
@@ -29,6 +30,38 @@ jobs:
...
+        - secure: HGKRJnJ2Y/JURFKWKWNtFpOMk1u99WIest+2SbuDBlDSVEtVWmUjUpAjz+TdaFt1XIaTANiQrHJdmsc4qgQW4QOhqjpOa4pm3CY74OfG7BYsBsaE5Ja8AiCYoiJ57VFejNQQdUMAzV8GWBjpX0CHbSwsG38ObyU6msql+DhLQXtyZZgbVXHq1DwrCAFB3xDazzePzcRuVikIZ70RoM5qa7VLxfWhI1oqHp31il4x3lr+vPYjWDqNulzXXv+8mD9A/GpEtAlDRJ2u2kH/3lSVNkIncUGKbb4Z2OiNUA97V7COLa31lhuDGSTAxVbf9xM9nnEHz/np3z9IHBZMvlCOCIBSPwgClCkUqEp0efKt6OufuAV/TkyFi0Llp4jeaTGeRRfPYF6hAZBbAR+2mOlyom6L6MBw0jrdnJq7euqdS/k/CdxkyMaXGGjqridGRrsRwZ//2ULcXbMgmKehSaqez/1ss4UtnV6LQqqbQQbZQppywz/j9bUoS8XGxaOAfhATRxOLbqQgKYmRD5YnJB+leVwFg3/YpTAiReEMrSwbyEN8OcAenHZtdCg+JwV/ZgobcerBTK0vXcWWlQgttqQEyT41qRU5kvkDfCc1UFDYWkVfaaRs0t+l4W+2YmtzIpDQk081JqFZHz6fXbadAIQu+J8Igap6bp9Y86eMUlLIwc4=
+      install:
+        - curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash
+      before_script:
+        - az login --service-principal --username "$USER_NAME"  --password "$PASSWORD" --tenant "$TENANT" > /dev/null
+        - ./buildZip.sh
```
Comment (kamiljsokolowski) [inline .travis.yml:55]: I can see that you are building the same `publish.zip` package twice during end-to-end and during the deployment phase.
We should be building a given artifact once, presumably during a dedicated build stage or along with unit testing, and utilizing it across all stages. Re-building is
a) redundant
b) can lead to config drifts (Travis destroy and re-deploys the build env after each stage)

File: .travis.yml [hunk 4]
```diff
@@ -29,6 +30,38 @@ jobs:
...
+        - az storage fs create --name "$STORAGE_ACCOUNT_DIR_NAME" --account-name "$STORAGE_ACCOUNT_NAME" --account-key "$STORAGE_ACCOUNT_KEY" --public-access file
+        - az storage fs file upload --file-system "$STORAGE_ACCOUNT_DIR_NAME" --account-name "$STORAGE_ACCOUNT_NAME" --account-key "$STORAGE_ACCOUNT_KEY" --source "publish.zip" --path "dynatrace-azure-log-forwarder.zip"
+        - az storage fs file upload --file-system "$STORAGE_ACCOUNT_DIR_NAME" --account-name "$STORAGE_ACCOUNT_NAME" --account-key "$STORAGE_ACCOUNT_KEY" --source "deployment/dynatrace-azure-forwarder.json" --path "dynatrace-azure-forwarder.json"
+        - az group create --name "$RESOURCE_GROUP_NAME" --location westeurope
+      script:
+        - ./deployment/dynatrace-azure-logs.sh --deployment-name "e2e-${TRAVIS_BUILD_ID}" --resource-group "$RESOURCE_GROUP_NAME" --target-url "$TARGET_URL" --target-api-token "$TARGET_API_TOKEN" --target-paas-token "$TARGET_PAAS_TOKEN" --event-hub-name "logs-forwarding-e2e-eh" --event-hub-connection-string "$EH_CONNECTION_STRING" --repository-release-url "https://${STORAGE_ACCOUNT_NAME}.blob.core.windows.net/${STORAGE_ACCOUNT_DIR_NAME}/"
```
Comment (kamiljsokolowski) [inline .travis.yml:61]: If the build artifact was shared between stages, you would not need the additional option to define other package sources.

File: .travis.yml (hunk 3 above)
Comment (Dawid-Kaszubski) [reply .travis.yml:55]: I've tried solving it using a sync solution analogical as for AWS S3. Unfortunately, Travis doesn't support Azure Storage Account, and installing aws cli only for S3 sync seems excessive.  
Also, I would like to keep creating and deleting Storage Account's directory in one stage. 
If you don't mind I will go with the current solution.

File: .travis.yml (hunk 4 above)
Comment (Dawid-Kaszubski) [reply .travis.yml:61]: For Azure deployment script is impossible. The Azure CLI command we use to create deployment requires ARM json from some URL. 
I don't want to change the deployment script only for E2E test.

**Review by sq2gxo — APPROVED** [review]

**Review by mswiatkowska — APPROVED** [review]

**Pull request merged**